### ISSUE TITLE
Identifier completion with syntax keywords

### ISSUE THREAD

**Issue by ptrv** [abingham/emacs-ycmd#258]
`ycmd` supports the feature to seed the identifier completion with syntax keywords.

@abingham Should we add this feature?

I have a branch with this feature implemented. Here is the diff: https://github.com/abingham/emacs-ycmd/compare/master...ptrv:syntax-keywords-support

The thing is that we need to supply the server with keywords and I haven't found a way to get the keywords from the mode, so I added a variable `ycmd-get-syntax-keywords-function`. The user needs to implement this to provide the keywords.

`company-mode` supports already a completion backend for keywords, but I think it is nicer if `ycmd` provides the keyword candidates itself instead of chaining completion backends.

Regarding `ycmd-get-syntax-keywords-function` it could be implemented the following ways:

- Reusing the variable `company-keyword-alist:

```lisp
(setq ycmd-syntax-keyword-function
      (lambda (mode)
          (assoc-default mode company-keyword-alist))
```

- or loading keywods from files:

```lisp
(setq ycmd-syntax-keyword-function
      (lambda (mode)
        (let ((filename (f-join "~/src/auto-complete/dict" (symbol-name mode))))
          (when (f-exists? filename)
            (let (result)
              (ignore-errors
                (with-temp-buffer
                  (insert-file-contents filename)
                  (setq result (s-split "\n" (buffer-string) t))))
              result)))))
```

**Comment by abingham** [comment]
I like this feature. It would be really useful for working with languages that don't have first-class ycmd support. We could even consider shipping pre-made syntax lists for languages with the client...we could let users supply them and we could just bundle them up (perhaps in a separate, related package). If you think this feature is ready to go - and it certainly looks fine to me - then let's merge it in.

**Comment by abingham** [comment]
On a related topic, we might think about encapsulating some of these "extra" features in their own packages. Features like "force-semantic", "tags", and this one all follow some similar patterns: they're optional, they add information to the "standard content", etc. It might make sense to split them out, if practical, into their own feature-specific files. 

This isn't a fully-formed plan on my part, just an observation. A full-blown plugin system might be over-engineering, but it might be the way to go if we get many more of these kinds of optional features. What do you think?

**Comment by ptrv** [comment]
Ok, I'll prepare the PR. I think the feature is ready to go, but still needs some documentation. I guess we can merge it as soon I added the info how to specify the `get-syntax-keywords-function` to the readme.

We could ship pre-made lists, like the package `auto-complete` is doing. At some point on my branch I was adding those too but removed them because I wasn't sure.

About the splitting of features, I am not sure, on the one hand it would improve readability I guess on the other it is still nice nice to have only one file because it is a self contained file. I personally did not feel the need to split out the parts while developing. I was also looking at other excellent packages like `flycheck`, they also have only one file.

**Comment by abingham** [comment]
Sounds good. I've been poking around with elm lately, so I'll be happy to
add an elm syntax file.

Regarding splitting out extra features, let's just keep things as they are
for now. There's definitely no strong need to make that kind of change
right now, and if you're happy with the state of things then we should just
keep them as they are. If the time comes that we really need to do
refactoring like that, we'll probably know it :)

**Comment by ptrv** [comment]
Am 18.01.2016 09:59 schrieb abingham:
> Sounds good. I've been poking around with elm lately, so I'll be happy
> to
>  add an elm syntax file.
Ok, then I'll add loading keyword files to the code in the PR

> 
>  Regarding splitting out extra features, let's just keep things as
> they are
>  for now. There's definitely no strong need to make that kind of
> change
>  right now, and if you're happy with the state of things then we
> should just
>  keep them as they are. If the time comes that we really need to do
>  refactoring like that, we'll probably know it :)
Sounds good.

**Comment by ptrv** [comment]
Resolved with #259
